### DISCUSSION TITLE
Please help me! In resizing QRCode TBitmap / Timage (I'm updating DelphiZXingQRCode opensource project)

### DISCUSSION THREAD

**Discussion by War3Evo** [skia4delphi/skia4delphi#162] [General]
I've been working on the opensource DelphiZXingQRCode to bring it up to speed with all the bug fixes via pull requests and I've made a threaded component for it.

I'm trying to figure a way to use Skia4Delphi so that I can resize the QRcode bitmap file without any blurriness and I've had no success.

I know I can load the QRcode into a TImage and resize it on the canvas on the form in windows 32bit... no blurriness at all at 2048 x 2048....but if I save the image, it saves as 32 by 32 pixels.  ??????

I also found out, I'll need to find a way to do this via FMX or Skia because the DLL that does it in windows will not work on other platforms.

Is there a way to code Skia4Delphi so I can add it to the opensource project that will Rescale or Resize the Bitmap image to just about any size?  **I will include your name in the credits of project gtDelphiZXingQRCode!** or you can update/pull request it, if you want.

I think if the Bitmap converts to a SVG, then resize the SVG to what ever is needed, then save the SVG into any image file would probably work better than just resizing a raster based file.  OR  I could write the pixel data into the SVG file, resize the SVG, then save the SVG as any image file (bmp, png, jpg, etc)??

[https://github.com/War3Evo/gtDelphi/tree/develop/gtDelphiZXingQRCode](https://github.com/War3Evo/gtDelphi/tree/develop/gtDelphiZXingQRCode)

Please help!

**Comment by viniciusfbb** [comment]
Hi!

Definitely the best way is to generate the SVG. But I don't agree to turn SVG into image after generation, it makes no sense. SVG works on everything, it's much smaller and the quality is 100% no matter what device resolution, resizing, zooming, etc.

In our source there is a demo where we generate a modern (rounded) QRCode based on Telegram's QRCodes design. In the code we generate the SVG and display the SVG and at no time do we convert the SVG to an image.

![Inspired_by_Telegram](https://user-images.githubusercontent.com/11139086/194716375-60829c55-d1d9-4d78-88bc-362e610dba79.png)

Source: [skia4delphi/Samples/Webinar - Demo/FMX/](https://github.com/skia4delphi/skia4delphi/tree/main/Samples/Webinar%20-%20Demo/FMX)
Video: [Skia4Delphi - Webinar Demo](https://www.youtube.com/watch?v=iKL8CvyFtk8)

**Reply by War3Evo** [reply]
Thank you so much for your quick reply!!

I looked at the rendering code and that's a lot of useful knowledge.  I will add credit and links to you for your help in this matter in the next update; Tomorrow.  I am doing daily updates.

I have another question, if you have time:

**There is code to load a SVG into Skia4Delphi via examples, but it says it doesn't support inner style??? So, can it support this style below???**

```
<rect
    width="1"
    height="1"
    x="10"
    y="10"
    style="fill: black; stroke: black; stroke-width: 1;" />
```

**Reply by viniciusfbb** [reply]
```xml
<rect
    width="1"
    height="1"
    x="10"
    y="10"
    fill=“black”
    stroke=“black”
    stroke-width=“1”
/>
```

When you are using image editors to made your svg (like photoshop), they have in theirs svg export options, a field with the option “embedded styles in attributes”, that does this job automatically for you.

**Reply by War3Evo** [reply]
I've update the source code.

I've added credits to you in the Readme.md and the commit.

It now has a optional Use Inner Style feature that is boolean.

I've made it very easy to change the pixel colors during the encoding process.

You can also create both BMP and SVG at the same time.

I'm next going to enable Skia4Delphi in a new Sample Project to see if I can get the SVG viewer working.   I wonder if Skia4Delphi can resize SVG and save it resized?

**Reply by viniciusfbb** [reply]
Yes, it can. But, it don’t make sense for me.

BTW, you can learn more about SVG generation with Skia4delphi in our SVG documentation: https://github.com/skia4delphi/skia4delphi/blob/main/Documents/SVG.md
